android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageFacebookConceal.java

```diff
@@ -0,0 +1,96 @@
+package com.oblador.keychain.cipherStorage;
+
+import android.os.Build;
+
+import com.facebook.android.crypto.keychain.AndroidConceal;
+import com.facebook.android.crypto.keychain.SharedPrefsBackedKeyChain;
+import com.facebook.crypto.Crypto;
+import com.facebook.crypto.CryptoConfig;
+import com.facebook.crypto.Entity;
+import com.facebook.crypto.keychain.KeyChain;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.oblador.keychain.exceptions.CryptoFailedException;
+
+import java.nio.charset.Charset;
+
+public class CipherStorageFacebookConceal implements CipherStorage {
+    public static final String CIPHER_STORAGE_NAME = "FacebookConceal";
+    public static final String KEYCHAIN_DATA = "RN_KEYCHAIN";
+    public static final String EMPTY_STRING = "";
+    private final Crypto crypto;
+
+    public CipherStorageFacebookConceal(ReactApplicationContext reactContext) {
+        KeyChain keyChain = new SharedPrefsBackedKeyChain(reactContext, CryptoConfig.KEY_256);
+        this.crypto = AndroidConceal.get().createDefaultCrypto(keyChain);
+    }
+
+    @Override
+    public String getCipherStorageName() {
+        return CIPHER_STORAGE_NAME;
+    }
+
+    @Override
+    public int getMinSupportedApiLevel() {
+        return Build.VERSION_CODES.JELLY_BEAN;
+    }
+
+    @Override
+    public EncryptionResult encrypt(String service, String username, String password) throws CryptoFailedException {
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+        service = service == null ? EMPTY_STRING : service;
+
+        Entity usernameEntity = createUsernameEntity(service);
+        Entity passwordEntity = createPasswordEntity(service);
+
+        try {
+            byte[] encryptedUsername = crypto.encrypt(username.getBytes(Charset.forName("UTF-8")), usernameEntity);
+            byte[] encryptedPassword = crypto.encrypt(password.getBytes(Charset.forName("UTF-8")), passwordEntity);
+
+            return new EncryptionResult(encryptedUsername, encryptedPassword, this);
+        } catch (Exception e) {
+            throw new CryptoFailedException("Encryption failed for service " + service, e);
+        }
+    }
+
+    @Override
+    public DecryptionResult decrypt(String service, byte[] username, byte[] password) throws CryptoFailedException {
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+        service = service == null ? EMPTY_STRING : service;
```

Comment:
I agree that there is a lot of copy-pasting regarding the null-check of service.

Defaulting to the empty string is used both in the `PrefsStorage` (because that's the way it used to be) and in the `CipherStorageFacebookConceal` (for the same reason), but the empty string cannot be used in the `CipherStorageFacebookAESCBC` because it is not allowed to use empty strings as keys in the Keystore. That is why I did not keep the null-check in the `KeychainModule` as it was before.

I could add some private methods in both `CipherStorageFacebookConceal` and `CipherStorageFacebookAESCBC` that does the null-check and defaulting..?